wal/wal.go

```diff
@@ -532,21 +540,28 @@ func (w *WAL) saveState(s *raftpb.HardState) error {
 }
 
 func (w *WAL) Save(st raftpb.HardState, ents []raftpb.Entry) error {
-	w.mu.Lock()
```

Comment:
some thoughts:
 - protobuf compiler generated method "func (m *Entry) MarshalTo(data []byte) (int, error)" needs a buffer sized Entry.Size() or more. this method will not append to the buffer
 - the way fmt package uses sync.Pool cannot be used here. fmt creates a handle which holds the buffer (initially unallocated) and uses append to grow the buffer. the generated code in raft.pb.go does not use this approach
 - we could use a simple approach of using sync.Pool with initial size of 1024. if this is not sufficient, we can allocate a new buf, use it and push to the Pool. i have added a snippet here : https://play.golang.org/p/fLKjslWKd6
 - saving this allocation is not enough, as "func (e *encoder) encode(rec *walpb.Record)" is going to marshal the Record, which is going to allocate again. only difference is Record will also contain the type. it is better we allocate for the Record and use MarshalTo() using the slices in it for Data/MetaData/Snapshot/Crc etc.